File: leetcode/problems/go/3020. Find the Maximum Number of Elements in Subset/solve.go
package main

import "slices"

func maximumLength(nums []int) int {
	if len(nums) == 0 {
		return 0
	}

	counts := make(map[int]int)
	for _, num := range nums {
		counts[num]++
	}
	checked := make(map[int]bool)
	slices.Sort(nums)
	maxLen := 1
	for _, num := range nums {
		if checked[num] {
			continue
		}
		checked[num] = true

		if num == 1 {
			if counts[num]%2 == 1 {
				maxLen = max(maxLen, counts[num])
			} else {
				maxLen = max(maxLen, counts[num]-1)
			}
			continue
		}

		curr := num
		currLen := 1
		for {
			if counts[curr] < 2 {
				break
			}
			next := curr * curr
			checked[next] = true
			if counts[next] == 0 {
				break
			}
			curr = next
			currLen += 2
			maxLen = max(maxLen, currLen)
		}
	}
	return maxLen
}

File: leetcode/problems/python3/3020. Find the Maximum Number of Elements in Subset/solve.py
from collections import Counter
from typing import List


class Solution:
    def maximumLength(self, nums: List[int]) -> int:
        if len(nums) == 0:
            return 0

        counts = Counter(nums)
        checked = set()
        nums = sorted(counts.keys())
        max_len = 1

        for num in nums:
            if num in checked:
                continue
            checked.add(num)

            if num == 1:
                max_len = max(max_len, counts[num] if counts[num] % 2 == 1 else counts[num] - 1)
                continue

            curr = num
            curr_len = 1
            while True:
                if counts[curr] < 2:
                    break
                next = curr**2
                checked.add(next)
                if not counts[next]:
                    break
                curr = curr**2
                curr_len += 2
                max_len = max(max_len, curr_len)

        return max_len
